util/conditions/experimental/getter.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	value, _, _ := unstructured.NestedFieldNoCopy(u.UnstructuredContent(), "status", "experimentalConditions")
+	if conditions, ok := value.([]interface{}); ok {
+		return convertUnstructuredConditions(conditions), nil
+	}
+
+	value, _, _ = unstructured.NestedFieldNoCopy(u.UnstructuredContent(), "status", "conditions")
+	if conditions, ok := value.([]interface{}); ok {
+		return convertUnstructuredConditions(conditions), nil
+	}
+
+	return nil, errors.New("obj must have Status with one of Conditions or ExperimentalConditions")
+}
+
+func convertUnstructuredConditions(conditions []interface{}) []metav1.Condition {
+	if conditions == nil {
+		return nil
+	}
+
+	convertedConditions := make([]metav1.Condition, 0, len(conditions))
+	for _, c := range conditions {
+		cMap, ok := c.(map[string]interface{})
+		if !ok || cMap == nil {
+			// TODO: think about returning an error in this case and when type of status are not set (as a signal it is a condition type)
+			continue
+		}
+
+		var conditionType string
+		if v, ok := cMap["type"]; ok {
+			conditionType = v.(string)
+		}
+
+		var status string
+		if v, ok := cMap["status"]; ok {
+			status = v.(string)
+		}
+
+		var observedGeneration int64
+		if v, ok := cMap["observedGeneration"]; ok {
+			observedGeneration = v.(int64)
+		}
+
+		var lastTransitionTime metav1.Time
+		if v, ok := cMap["lastTransitionTime"]; ok {
+			_ = lastTransitionTime.UnmarshalQueryParameter(v.(string))
+		}
+
+		var reason string
+		if v, ok := cMap["reason"]; ok {
+			reason = v.(string)
+		}
+
+		var message string
+		if v, ok := cMap["message"]; ok {
+			message = v.(string)
+		}
+
+		convertedConditions = append(convertedConditions, metav1.Condition{
+			Type:               conditionType,
+			Status:             metav1.ConditionStatus(status),
+			ObservedGeneration: observedGeneration,
+			LastTransitionTime: lastTransitionTime,
+			Reason:             reason,
+			Message:            message,
+		})
+	}
+	return convertedConditions
+}
+
+func getFromTypedObject(obj runtime.Object) ([]metav1.Condition, error) {
+	ptr := reflect.ValueOf(obj)
+	if ptr.Kind() != reflect.Pointer {
+		return nil, errors.New("obj must be a pointer")
+	}
+
+	elem := ptr.Elem()
+	if !elem.IsValid() {
+		return nil, errors.New("obj must be a valid value (non zero value of its type)")
+	}
+
+	statusField := elem.FieldByName("Status")
+	if statusField == (reflect.Value{}) {
+		return nil, errors.New("obj must have a Status field")
+	}
+
+	// Get conditions.
+	// NOTE: Given that we allow providers to migrate at different speed, it is required to support objects at the different stage of the transition from legacy conditions to metav1.conditions.
+	// In order to handle this, first try to read Status.ExperimentalConditions, then Status.Conditions; for Status.Conditions, also support conversion from legacy conditions.
+	// The ExperimentalConditions branch and the conversion from legacy conditions should be dropped when v1beta1 API are removed.
```

Comment:
When we deal with providers objects, they are in the for of unstructured objects, and thus getFromUnstructuredObject is used instead of this function (and I agree we should keep handling legacy conditions there).

In this func instead I think we can(and should drop) as soon as possible.

Reply:
(sorry might be wrong, would have to check the code again)

> they are in the for of unstructured objects

That is only true if the utils are used in core CAPI, right? If the util is used in provider code they have the type?

Reply:
Discussed, we can resolve